File: user-service/src/main/java/com/bank/userservice/security/config/KeycloakJwtAuthenticationConverter.java
package com.bank.userservice.security.config;

import com.bank.userservice.security.CurrentUser;
import lombok.NonNull;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.core.convert.converter.Converter;
import org.springframework.security.authentication.AbstractAuthenticationToken;
import org.springframework.security.core.GrantedAuthority;
import org.springframework.security.core.authority.SimpleGrantedAuthority;
import org.springframework.security.oauth2.core.OAuth2AccessToken;
import org.springframework.security.oauth2.jwt.Jwt;
import org.springframework.security.oauth2.server.resource.authentication.BearerTokenAuthentication;
import org.springframework.security.oauth2.server.resource.authentication.JwtGrantedAuthoritiesConverter;
import org.springframework.stereotype.Component;

import java.util.Collection;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.UUID;
import java.util.stream.Collectors;
import java.util.stream.Stream;

@Component
public class KeycloakJwtAuthenticationConverter implements Converter<Jwt, AbstractAuthenticationToken> {

    private final JwtGrantedAuthoritiesConverter defaultGrantedAuthoritiesConverter =
            new JwtGrantedAuthoritiesConverter();

    @Value("${keycloak.resource}")
    private String clientId;

    @Override
    @NonNull
    public AbstractAuthenticationToken convert(@NonNull Jwt jwt) {
        Collection<GrantedAuthority> defaultAuthorities = defaultGrantedAuthoritiesConverter.convert(jwt);
        Collection<GrantedAuthority> authorities = Stream.concat(
                defaultAuthorities.stream(),
                extractKeycloakAuthorities(jwt).stream()
        ).collect(Collectors.toSet());

        UUID userId = UUID.fromString(jwt.getSubject());
        String email = jwt.getClaimAsString("preferred_username");
        Map<String, Object> attributes = jwt.getClaims();

        CurrentUser currentUser = new CurrentUser(userId, email, attributes, authorities);

        OAuth2AccessToken accessToken = new OAuth2AccessToken(
                OAuth2AccessToken.TokenType.BEARER,
                jwt.getTokenValue(),
                jwt.getIssuedAt(),
                jwt.getExpiresAt()
        );

        return new BearerTokenAuthentication(currentUser, accessToken, authorities);
    }


    @SuppressWarnings("unchecked")
    private Collection<GrantedAuthority> extractKeycloakAuthorities(Jwt jwt) {
        Set<GrantedAuthority> authorities = new HashSet<>();

        Map<String, Object> realmAccess = jwt.getClaim("realm_access");
        if (realmAccess != null && realmAccess.containsKey("roles")) {
            authorities.addAll(((List<String>) realmAccess.get("roles")).stream()
                    .map(role -> "ROLE_" + role.toUpperCase())
                    .map(SimpleGrantedAuthority::new)
                    .collect(Collectors.toSet()));
        }

        Map<String, Object> resourceAccess = jwt.getClaimAsMap("resource_access");
        if (resourceAccess != null && resourceAccess.containsKey(clientId)) {
            Map<String, Object> resource = (Map<String, Object>) resourceAccess.get(clientId);
            authorities.addAll(((List<String>) resource.get("roles")).stream()
                    .map(role -> "ROLE_" + role.toUpperCase())
                    .map(SimpleGrantedAuthority::new)
                    .collect(Collectors.toSet()));
        }

        return authorities;
    }
}

File: core-query/src/main/kotlin/ru/patterns/corequery/security/config/KeycloakJwtAuthenticationConverter.kt
package ru.patterns.corequery.security.config

import org.springframework.core.convert.converter.Converter
import org.springframework.security.authentication.AbstractAuthenticationToken
import org.springframework.security.core.GrantedAuthority
import org.springframework.security.core.authority.SimpleGrantedAuthority
import org.springframework.security.oauth2.core.OAuth2AccessToken
import org.springframework.security.oauth2.jwt.Jwt
import org.springframework.security.oauth2.server.resource.authentication.BearerTokenAuthentication
import org.springframework.security.oauth2.server.resource.authentication.JwtGrantedAuthoritiesConverter
import org.springframework.stereotype.Component
import reactor.core.publisher.Mono
import reactor.kotlin.core.publisher.toMono
import ru.patterns.core.security.CurrentUser
import java.util.Locale
import java.util.UUID
import java.util.stream.Collectors
import java.util.stream.Stream


@Component
class KeycloakJwtAuthenticationConverter : Converter<Jwt, Mono<AbstractAuthenticationToken>> {
    private val defaultGrantedAuthoritiesConverter = JwtGrantedAuthoritiesConverter()

    override fun convert(jwt: Jwt): Mono<AbstractAuthenticationToken> {
        val defaultAuthorities = defaultGrantedAuthoritiesConverter.convert(jwt)!!
        val authorities: Collection<GrantedAuthority> = Stream.concat(
            defaultAuthorities.stream(),
            extractKeycloakAuthorities(jwt).stream()
        ).collect(Collectors.toSet())

        val userId = UUID.fromString(jwt.subject)
        val email = jwt.getClaimAsString("preferred_username")
        val attributes = jwt.claims

        val currentUser = CurrentUser(userId, email, attributes, authorities)

        val accessToken = OAuth2AccessToken(
            OAuth2AccessToken.TokenType.BEARER,
            jwt.tokenValue,
            jwt.issuedAt,
            jwt.expiresAt
        )

        return BearerTokenAuthentication(currentUser, accessToken, authorities).toMono()
    }


    private fun extractKeycloakAuthorities(jwt: Jwt): Collection<GrantedAuthority> {
        val authorities: MutableSet<GrantedAuthority> = HashSet()

        val realmAccess = jwt.getClaim<Map<String, Any>>("realm_access")
        if (realmAccess != null && realmAccess.containsKey("roles")) {
            authorities.addAll(
                (realmAccess["roles"] as List<String>)
                    .map { role -> "ROLE_" + role.uppercase(Locale.getDefault()) }
                    .map { role -> SimpleGrantedAuthority(role) }
                    .toSet())
        }

        return authorities
    }
}